File: ui/frontend/src/api/changelog.ts
import { apiCall } from './client';

export interface ChangelogGroup {
  title: string;
  items: string[];
}

export interface ChangelogEntry {
  version: string;
  date: string;
  groups: ChangelogGroup[];
}

export interface ChangelogResult {
  available: boolean;
  current: string;
  last_seen?: string;
  markdown: string;
  entries: ChangelogEntry[];
}

// Paths are relative to config.apiUrl (which already includes "/api"); do NOT
// prepend "/api" here.
export const changelogApi = {
  unseen: (): Promise<{ has_unseen: boolean }> =>
    apiCall('/changelog/unseen').then((r) => r.json()),
  get: (params?: { all?: boolean; since?: string }): Promise<ChangelogResult> => {
    const qs = new URLSearchParams();
    if (params?.since) qs.set('since', params.since);
    else if (params?.all) qs.set('range', 'all');
    const suffix = qs.toString() ? `?${qs.toString()}` : '';
    return apiCall(`/changelog${suffix}`).then((r) => r.json());
  },
};

File: internal/cliclient/changelog.go
package cliclient

import (
	"net/http"
	"net/url"
)

// ChangelogGroup is a titled list of change descriptions.
type ChangelogGroup struct {
	Title string   `json:"title"`
	Items []string `json:"items"`
}

// ChangelogEntry is one released version's notes.
type ChangelogEntry struct {
	Version string           `json:"version"`
	Date    string           `json:"date"`
	Groups  []ChangelogGroup `json:"groups"`
}

// ChangelogResult is the GET /api/changelog response.
type ChangelogResult struct {
	Available bool             `json:"available"`
	Current   string           `json:"current"`
	LastSeen  string           `json:"last_seen"`
	Markdown  string           `json:"markdown"`
	Entries   []ChangelogEntry `json:"entries"`
}

// GetChangelog fetches the changelog. With rangeAll it requests the full
// history (auto-marks seen server-side); with a non-empty since it requests a
// pure read of entries newer than that version; otherwise it requests the
// since-last diff (auto-marks seen). rangeAll and since are mutually exclusive;
// since takes precedence if both are set.
func (c *Client) GetChangelog(key string, rangeAll bool, since string) (*ChangelogResult, error) {
	q := url.Values{}
	switch {
	case since != "":
		q.Set("since", since)
	case rangeAll:
		q.Set("range", "all")
	}
	path := "/api/changelog"
	if enc := q.Encode(); enc != "" {
		path += "?" + enc
	}
	var out ChangelogResult
	if err := c.doBearer(http.MethodGet, path, key, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
